File: davidbuccola-commons-spark/src/main/java/net/davidbuccola/commons/spark/AvroFileInputFormat.java
package net.davidbuccola.commons.spark;

import org.apache.avro.Schema;
import org.apache.avro.file.DataFileReader;
import org.apache.avro.file.SeekableInput;
import org.apache.avro.io.DatumReader;
import org.apache.avro.mapreduce.AvroRecordReaderBase;
import org.apache.avro.specific.SpecificRecord;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.mapreduce.InputSplit;
import org.apache.hadoop.mapreduce.RecordReader;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;

import java.io.IOException;

/**
 * A base class for AVRO {@link FileInputFormat} implementations which have a couple important characteristics useful
 * for Spark streaming:
 * <p>
 * 1) A new {@link T} instance is created for each datum (as opposed to reusing the instance). This means the {@link T}
 * instance can be held for a while without fear of being overwritten. This is particularly useful when sets of records
 * are gathered up for processing further down the Spark pipeline.
 * <p>
 * 2) The schema is obtained from the datum class rather than the Hadoop configuration.
 */
public abstract class AvroFileInputFormat<T extends SpecificRecord> extends FileInputFormat<T, NullWritable> {

    private final Schema schema;

    protected AvroFileInputFormat(Class<T> datumClass) {
        this.schema = getSchema(datumClass);
    }

    @Override
    public RecordReader<T, NullWritable> createRecordReader(InputSplit split, TaskAttemptContext context) {
        return new AvroRecordReader();
    }

    private Schema getSchema(Class<T> datumClass) {
        try {
            return (Schema) datumClass.getDeclaredMethod("getClassSchema").invoke(null);
        } catch (Exception e) {
            throw new IllegalArgumentException("Failed to get schema for datum class", e);
        }
    }

    private class AvroRecordReader extends AvroRecordReaderBase<T, NullWritable, T> {

        AvroRecordReader() {
            super(schema);
        }

        @Override
        public T getCurrentKey() throws IOException, InterruptedException {
            return getCurrentRecord();
        }

        @Override
        public NullWritable getCurrentValue() throws IOException, InterruptedException {
            return NullWritable.get();
        }

        @Override
        protected DataFileReader<T> createAvroFileReader(SeekableInput input, DatumReader<T> datumReader) throws IOException {
            return new DataFileReader<T>(input, datumReader) {
                @Override
                public T next(T reuse) throws IOException {
                    return super.next(null); // Don't reuse the buffer even though the caller wanted us to
                }
            };
        }
    }
}

File: davidbuccola-commons-spark/src/main/scala/net/davidbuccola/commons/spark/AvroFileInputFormat.scala
package net.davidbuccola.commons.spark

import org.apache.avro.Schema
import org.apache.avro.file.{DataFileReader, SeekableInput}
import org.apache.avro.io.DatumReader
import org.apache.avro.mapreduce.AvroRecordReaderBase
import org.apache.avro.specific.SpecificRecord
import org.apache.hadoop.io.NullWritable
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
import org.apache.hadoop.mapreduce.{InputSplit, TaskAttemptContext}

import scala.reflect.{ClassTag, classTag}

/**
  * A [[FileInputFormat]] for AVRO which has a couple important characteristics useful
  * for Spark streaming:
  * <p>
  * 1) A new [[T]] instance is created for each datum (as opposed to reusing the instance). This means the [[T]]
  * instance can be held for a while without fear of being overwritten. This is particularly useful when sets of records
  * are gathered up for processing further down the Spark pipeline.
  * <p>
  * 2) The schema is obtained from the datum class rather than the Hadoop configuration.
  */
class AvroFileInputFormat[T <: SpecificRecord : ClassTag] extends FileInputFormat[T, NullWritable] {

  private val datumClass = classTag[T].runtimeClass.asInstanceOf[Class[T]]
  private val schema =
    try
      datumClass.getDeclaredMethod("getClassSchema").invoke(null).asInstanceOf[Schema]
    catch {
      case e: Exception =>
        throw new IllegalArgumentException("Failed to get schema for datum class", e)
    }

  override def createRecordReader(split: InputSplit, context: TaskAttemptContext) = new AvroRecordReader

  class AvroRecordReader extends AvroRecordReaderBase[T, NullWritable, T](schema) {

    override def getCurrentKey: T = getCurrentRecord

    override def getCurrentValue: NullWritable = NullWritable.get

    override def createAvroFileReader(input: SeekableInput, datumReader: DatumReader[T]): DataFileReader[T] = new DataFileReader[T](input, datumReader) {

      override def next(reuse: T): T = super.next(null.asInstanceOf[T]) // Don't reuse the buffer even though the caller wanted us to
    }
  }

}
